File: backend/app/schemas/billing.py
"""비용/Billing 스키마 (A-BL-01·02).

키는 요청으로만 들어오고 응답 모델에는 **존재하지 않는다** — 실수로 흘리지 않도록
아예 필드를 두지 않는다(클러스터 자격증명과 같은 방식).
"""

from datetime import datetime

from pydantic import BaseModel, Field


class BillingCredentialsPut(BaseModel):
    access_key: str = Field(..., min_length=8)
    secret_key: str = Field(..., min_length=8)
    monthly_alert_krw: int | None = Field(default=None, ge=0)


class BillingConfigOut(BaseModel):
    configured: bool
    scp_account_id: str | None = None
    api_endpoint: str | None = None
    monthly_alert_krw: int | None = None
    last_verified_at: datetime | None = None


class BillingRuleIn(BaseModel):
    kind: str = Field(default="tag", pattern="^(tag|resource_id|billing_item_id|service_category)$")
    condition: str = Field(..., min_length=1, max_length=255)
    mapping_label: str | None = Field(default=None, max_length=128)

File: frontend/src/api/billing.ts
import { api } from './client'

export interface BillingConfig {
  configured: boolean
  scp_account_id: string | null
  api_endpoint: string | null
  monthly_alert_krw: number | null
  last_verified_at: string | null
}

export interface BillingTrend {
  start: string
  end: string
  total_krw: number
  daily: { date: string; krw: number }[]
  by_category: { label: string; krw: number }[]
  by_item: { label: string; krw: number }[]
  record_count: number
  tag: string | null
  tagged_resource_count: number | null
}

export interface BillingRule {
  id: number
  kind: string
  condition: string
  mapping_label: string | null
  is_active: boolean
}

export const billingApi = {
  config: () => api.get<BillingConfig>('/billing/config'),
  /** 키는 요청으로만 나가고 응답에는 실리지 않는다. 서버가 저장 전 실제 조회로 검증한다. */
  putConfig: (payload: {
    access_key: string
    secret_key: string
    monthly_alert_krw?: number | null
  }) => api.put<BillingConfig>('/billing/config', payload),
  trend: (days = 30, tag?: string) =>
    api.get<BillingTrend>(
      `/billing/trend?days=${days}${tag ? `&tag=${encodeURIComponent(tag)}` : ''}`,
    ),
  rules: () => api.get<BillingRule[]>('/billing/rules'),
  addRule: (payload: { kind: string; condition: string; mapping_label?: string }) =>
    api.post<BillingRule>('/billing/rules', payload),
  removeRule: (id: number) => api.del<{ ok: boolean }>(`/billing/rules/${id}`),
}
